source/Calamari.Terraform/TerraformCliExecutor.cs

```diff
@@ -0,0 +1,212 @@
+        readonly Dictionary<string, string> environmentVariables;
+        private Dictionary<string, string> defaultEnvironmentVariables;
+        private readonly string logPath;
+
+        public TerraformCLIExecutor(ICalamariFileSystem fileSystem, RunningDeployment deployment,
+            Dictionary<string, string> environmentVariables)
+        {
+            this.fileSystem = fileSystem;
+            this.deployment = deployment;
+            this.environmentVariables = environmentVariables;
+
+            var variables = deployment.Variables;
+            TemplateDirectory = variables.Get(TerraformSpecialVariables.Action.Terraform.TemplateDirectory, deployment.CurrentDirectory);
+            Workspace = variables.Get(TerraformSpecialVariables.Action.Terraform.Workspace);
+            InitParams = variables.Get(TerraformSpecialVariables.Action.Terraform.AdditionalInitParams);
+            ActionParams = variables.Get(TerraformSpecialVariables.Action.Terraform.AdditionalActionParams);
+            TerraformExecutable = variables.Get(TerraformSpecialVariables.Action.Terraform.CustomTerraformExecutable) ??
+                                  $"terraform{(CalamariEnvironment.IsRunningOnWindows ? ".exe" : String.Empty)}";
+            AllowPluginDownloads = variables.GetFlag(TerraformSpecialVariables.Action.Terraform.AllowPluginDownloads, true);
+            AttachLogFile = variables.GetFlag(TerraformSpecialVariables.Action.Terraform.AttachLogFile);
+            TerraformVariableFiles = GenerateVarFiles();
+
+            logPath = Path.Combine(deployment.CurrentDirectory, "terraform.log");
+
+            if (!String.IsNullOrEmpty(TemplateDirectory))
+            {
+                var templateDirectoryTemp = Path.Combine(deployment.CurrentDirectory, TemplateDirectory);
+
+                if (!Directory.Exists(templateDirectoryTemp))
+                {
+                    throw new Exception($"Directory {TemplateDirectory} does not exist.");
+                }
+
+                TemplateDirectory = templateDirectoryTemp;
+            }
+
+            InitialiseTerraformEnvironmentVariables();
+
+            InitialisePlugins();
+
+            InitialiseWorkspace();
+        }
+
+        public string ExecuteCommand(params string[] arguments)
+        {
+            var commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out var result);
+
+            commandResult.VerifySuccess();
+
+            return result;
+        }
+
+        public int ExecuteCommand(out string result, params string[] arguments)
+        {
+            var commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out result);
+
+            return commandResult.ExitCode;
+        }
+
+        public int ExecuteCommand(out string result, out CommandResult commandResult, params string[] arguments)
+        {
+            commandResult = ExecuteCommandInternal(ToSpaceSeparated(arguments), out result);
+
+            return commandResult.ExitCode;
```

Comment:
rather than using an `out` parameter, then returning a value that comes off that same object, can we just set the return type to be `CommandResult` then consuming code can do with it what it wants (as a plus, its more obvious that the returned value is actually the exit code, since if you need the value you would use something like `result.ExitCode`

Reply:
fixed